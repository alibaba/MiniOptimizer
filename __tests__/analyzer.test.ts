import path from 'path';
import Analyzer from '../src/analyzer';

jest.setTimeout(60 * 1e3);

function T(appname: string) {
  return path.resolve(__dirname, `../fixtures/${appname}`);
}

test('New Analyze with sub packages', async () => {
  const analyzer = new Analyzer(T('mini-app-3'), 'wx');
  await analyzer.analyze();
  const all_files = analyzer.all_files;
  const main_files = analyzer.get_main_files();
  expect(all_files.filter((item) => !item.startsWith('pages/home/')).sort()).toEqual(main_files.sort());
  const main_unused_files = analyzer.get_main_unused_files();
  const useless_files = analyzer.get_useless_files();
  expect(main_unused_files.sort()).toEqual(
    [
      'abc.json',
      'common/a.js',
      'assets/demo.png',
      'common/b.js',
      'common/c.js',
      'common/d.js',
      'components/com-a/header.wxml',
      'components/com-a/index.js',
      'components/com-a/index.json',
      'components/com-a/index.wxml',
      'components/com-a/index.wxss',
      'styles/b.wxss',
    ].sort()
  );
  expect(useless_files.sort()).toEqual(['abc.json', 'common/d.js', 'pages/home/xxx.js'].sort());
  const main_used_files = analyzer.get_main_used_files();
  expect(main_used_files.sort()).toEqual(
    [
      'pages/index/index.js',
      'pages/index/ab.js',
      'pages/index/index.json',
      'pages/index/index.wxml',
      'pages/index/header.wxml',
      'pages/index/templates.wxml',
      'pages/index/tool.wxs',
      'pages/index/index.wxss',
      'sitemap.json',
      'app.js',
      'common/config.js',
      'others/a.js',
      'others/b.js',
      'app.json',
      'assets/tab/home.png',
      'assets/tab/home_select.png',
      'assets/tab/demo.png',
      'assets/tab/demo_select.png',
      'app.wxss',
    ].sort()
  );
});

test('New Analyze only main', async () => {
  const analyzer = new Analyzer(T('mini-app-1'), 'wx');
  await analyzer.analyze();
  const all_files = analyzer.all_files;
  const main_files = analyzer.get_main_files();
  const main_used_files = analyzer.get_main_used_files();
  const main_unused_files = analyzer.get_main_unused_files();
  const useless_files = analyzer.get_useless_files();
  expect(all_files).toMatchSnapshot();
  expect(main_files).toMatchSnapshot();
  expect(main_used_files).toMatchSnapshot();
  expect(main_unused_files).toMatchSnapshot();
  expect(useless_files).toMatchSnapshot();
});

test('get_sub_main_deps', async () => {
  const analyzer = new Analyzer(T('mini-app-5'), 'wx');
  await analyzer.analyze();
  const sub_deps = analyzer.get_sub_outside_files();
  expect(sub_deps).toHaveProperty('pages/home');
  expect(sub_deps['pages/home']).toHaveLength(10);
  expect(
    [
      'assets/demo.png',
      'common/a.js',
      'common/b.js',
      'common/c.js',
      'styles/b.wxss',
      'components/com-a/index.js',
      'components/com-a/index.wxml',
      'components/com-a/index.json',
      'components/com-a/index.wxss',
      'components/com-a/header.wxml',
    ].every((item) => sub_deps['pages/home'].includes(item))
  ).toBe(true);
});
