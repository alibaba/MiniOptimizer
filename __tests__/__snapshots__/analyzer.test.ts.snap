// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Analyze only main 1`] = `
Array [
  "abc.json",
  "app.js",
  "app.json",
  "app.wxss",
  "assets/demo.png",
  "common/a.js",
  "common/b.js",
  "common/c.js",
  "common/config.js",
  "common/d.js",
  "components/comp-a/index.js",
  "components/comp-a/index.json",
  "components/comp-a/index.wxml",
  "components/comp-a/index.wxss",
  "ext.json",
  "pages/home/header.wxml",
  "pages/home/index.js",
  "pages/home/index.json",
  "pages/home/index.wxml",
  "pages/home/index.wxss",
  "pages/home/templates.wxml",
  "pages/home/tool.wxs",
  "project.config.json",
  "sitemap.json",
  "styles/b.wxss",
]
`;

exports[`New Analyze only main 2`] = `
Array [
  "abc.json",
  "app.js",
  "app.json",
  "app.wxss",
  "assets/demo.png",
  "common/a.js",
  "common/b.js",
  "common/c.js",
  "common/config.js",
  "common/d.js",
  "components/comp-a/index.js",
  "components/comp-a/index.json",
  "components/comp-a/index.wxml",
  "components/comp-a/index.wxss",
  "ext.json",
  "pages/home/header.wxml",
  "pages/home/index.js",
  "pages/home/index.json",
  "pages/home/index.wxml",
  "pages/home/index.wxss",
  "pages/home/templates.wxml",
  "pages/home/tool.wxs",
  "project.config.json",
  "sitemap.json",
  "styles/b.wxss",
]
`;

exports[`New Analyze only main 3`] = `
Array [
  "app.js",
  "common/config.js",
  "others/a.js",
  "others/b.js",
  "app.json",
  "app.wxss",
  "pages/home/index.js",
  "common/a.js",
  "common/b.js",
  "common/c.js",
  "pages/home/index.json",
  "pages/home/index.wxml",
  "pages/home/header.wxml",
  "assets/demo.png",
  "pages/home/templates.wxml",
  "pages/home/tool.wxs",
  "pages/home/index.wxss",
  "styles/b.wxss",
  "sitemap.json",
  "ext.json",
  "project.config.json",
]
`;

exports[`New Analyze only main 4`] = `
Array [
  "abc.json",
  "common/d.js",
  "components/comp-a/index.js",
  "components/comp-a/index.json",
  "components/comp-a/index.wxml",
  "components/comp-a/index.wxss",
]
`;

exports[`New Analyze only main 5`] = `
Array [
  "abc.json",
  "common/d.js",
  "components/comp-a/index.js",
  "components/comp-a/index.json",
  "components/comp-a/index.wxml",
  "components/comp-a/index.wxss",
]
`;
